prophy/tests/test_array_bound.py

```diff
@@ -134,17 +136,20 @@ class XS(prophy.with_metaclass(prophy.struct_generator, prophy.struct_packed)):
     x.decode(b"\x06\x01\x02\x03\x04", ">")
     assert x.value[:] == [1, 2, 3, 4]
 
-    with pytest.raises(Exception) as e:
+    with pytest.raises(prophy.DecodeError) as e:
```

Comment:
Krzysztof, I don't think that anybody here is be able to separate 50 error classes. I've reached 5 and have a hard time to find out more. I also cannot understand why do you refuse using well known exception class' inheritance.
The thing bellow is my proposal for exception hierarchy.

```
ProphyError
  - DefinitionError - raised while interpreting given class' definition
  - DecodeError - raised under call of .decode() method
      * DataSizeError 
          + too few bytes to decode given type
          + no all bytes read
      * ConstraintError 
          + illegal value of a sizer, e.g. out of bound (in limitted array), negative size
          + unknown union's discriminator
          + illegal enumerator value
  - AssignmentError - raised while filling values
      + assigned type mismatch 
      + assigned slice size out of bound
      + exceeded array's size limit (e.g. too many .add() called  - it's not really an assignment so the exception name doe not fit here, but I cannot find any other)
```
I think there is no place for `EncodeError`, because all the validation happens while assignment / filling data.
![prophy_errors_proposal](https://user-images.githubusercontent.com/9801087/48056613-6fbebe80-e1b2-11e8-8574-38ecd2fa51fc.png)

Reply:
I will try to prepare second iteration of this PR by the end of this week

Reply:
>  could DataSizeError be renamed to a more general BufferError?
Sure but it still does not solve the original problem I try to resolve -- distinguishing at least the overflow error from the rest. I will probably discuss this matter with @kamichal offline and send another graph to continue the discussion here before writing any code because I think that some of the lower-level bullet points from his list deserve to get unique error classes. It's not clear to me at the moment whether we still want to use error codes or not.